imagetest/testworkflow.go

```diff
@@ -0,0 +1,245 @@
+package imagetest
+
+import (
+	"context"
+	"encoding/xml"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	"github.com/GoogleCloudPlatform/guest-test-infra/imagetest/utils"
+)
+
+var (
+	client *storage.Client
+)
+
+const (
+	testBinariesPath  = "/out"
+	testWrapperPath   = testBinariesPath + "/wrapper"
+	baseGCSPath       = "gs://gcp-guest-test-outputs/cloud_image_tests/"
+	createVMsStepName = "create-vms"
+)
+
+// TestWorkflow defines a test workflow which creates at least one test VM.
+type TestWorkflow struct {
+	Image          string
+	Name           string
+	ShortImage     string
+	destination    string
+	skipped        bool
+	skippedMessage string
+	wf             *daisy.Workflow
+}
+
+func finalizeWorkflows(tests []*TestWorkflow, zone, project string) error {
+	run := time.Now().Format(time.RFC3339)
+	for _, ts := range tests {
+		if ts.wf == nil {
+			continue
+		}
+		ts.destination = fmt.Sprintf("%s/%s/%s/%s", baseGCSPath, run, ts.Name, ts.ShortImage)
+
+		ts.wf.GCSPath = ts.destination
+
+		ts.wf.DisableGCSLogging()
+		ts.wf.DisableCloudLogging()
+		ts.wf.DisableStdoutLogging()
+
+		ts.wf.Zone = zone
+		ts.wf.Project = project
+
+		ts.wf.Sources["wrapper"] = testWrapperPath
+		ts.wf.Sources["testpackage"] = fmt.Sprintf("%s/%s.test", testBinariesPath, ts.Name)
+
+		copyGCSObject := daisy.CopyGCSObject{}
+		copyGCSObject.Source = "${OUTSPATH}/" // Trailing slash apparently crucial.
+		copyGCSObject.Destination = ts.destination + "/outs"
+		copyGCSObjects := &daisy.CopyGCSObjects{copyGCSObject}
+		copyStep, err := ts.wf.NewStep("copy-objects")
+		if err != nil {
+			return fmt.Errorf("failed to add copy-objects step to workflow %s: %v", ts.Name, err)
+		}
+		copyStep.CopyGCSObjects = copyGCSObjects
+
+		// The "copy-objects" step depends on every wait step.
+		for stepname, step := range ts.wf.Steps {
+			if !strings.HasPrefix(stepname, "wait-") {
+				continue
+			}
+			if err := ts.wf.AddDependency(copyStep, step); err != nil {
+				return fmt.Errorf("failed to add copy-objects step dependency to workflow %s: %v", ts.Name, err)
+			}
+		}
+
+	}
+	return nil
+}
+
+type testResult struct {
+	testWorkflow    *TestWorkflow
+	skipped         bool
+	workflowSuccess bool
+	err             error
+	results         []string
+}
+
+func getTestResults(ctx context.Context, ts *TestWorkflow) ([]string, error) {
```

Comment:
this can be a method on *TestWorkflow

Reply:
is it beneficial to make anything that can be a method into a method? i tend to make methods when there will be some shared data in the object that is going to be modified, or when i want the method to be public to the user of the package.

Reply:
The criteria whether something is a method that I use is whether something logically belongs to a class/struct or it operates mostly on class fields. For example, getTestResults retrieves results from a TestWorkflow. It mostly references TestWorkflow fields, so is a good candidate for a method.